src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/AbstractConverter.cs

```diff
@@ -0,0 +1,78 @@
+                    continuation: null)) // The current coverage of foreach statements to support does not need to use query continuations.                                                                                                           
+            .WithAdditionalAnnotations(Formatter.Annotation);
+
+        private static QueryClauseSyntax CreateQueryClause(ExtendedSyntaxNode node)
+        {
+            switch (node.Node.Kind())
+            {
+                case SyntaxKind.VariableDeclarator:
+                    var variable = (VariableDeclaratorSyntax)node.Node;
+                    return SyntaxFactory.LetClause(SyntaxFactory.Token(SyntaxKind.LetKeyword), variable.Identifier, variable.Initializer.EqualsToken, variable.Initializer.Value)
+                        .WithComments(node.ExtraLeadingComments, node.ExtraTrailingComments);
+
+                case SyntaxKind.ForEachStatement:
+                    return CreateFromClause((ForEachStatementSyntax)node.Node, node.ExtraLeadingComments, node.ExtraTrailingComments);
+
+                case SyntaxKind.IfStatement:
+                    var ifStatement = (IfStatementSyntax)node.Node;
+                    return SyntaxFactory.WhereClause(
+                                SyntaxFactory.Token(SyntaxKind.WhereKeyword).WithComments(ifStatement.IfKeyword.LeadingTrivia, ifStatement.IfKeyword.TrailingTrivia),
+                                ifStatement.Condition.WithComments(new[] { ifStatement.OpenParenToken }, new[] { ifStatement.CloseParenToken }))
+                                .WithComments(node.ExtraLeadingComments, node.ExtraTrailingComments);
+            }
+
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        private static FromClauseSyntax CreateFromClause(ForEachStatementSyntax forEachStatement, IEnumerable<SyntaxTrivia> extraLeadingTrivia, IEnumerable<SyntaxTrivia> extraTrailingTrivia)
+            => SyntaxFactory.FromClause(
+                    fromKeyword: SyntaxFactory.Token(SyntaxKind.FromKeyword)
+                        .WithComments(forEachStatement.ForEachKeyword.LeadingTrivia, forEachStatement.ForEachKeyword.TrailingTrivia, forEachStatement.OpenParenToken).KeepCommentsAndAddElasticMarkers(),
+                    type: forEachStatement.Type.IsVar ? null : forEachStatement.Type,
+                    identifier: forEachStatement.Type.IsVar ? forEachStatement.Identifier.WithCommentsBeforeLeadingTrivia(Helpers.GetTrivia(forEachStatement.Type.GetFirstToken())) : forEachStatement.Identifier,
+                    inKeyword: forEachStatement.InKeyword.KeepCommentsAndAddElasticMarkers(),
+                    expression: forEachStatement.Expression)
+                .WithComments(extraLeadingTrivia, extraTrailingTrivia, forEachStatement.CloseParenToken);
+
+        public abstract void Convert(SyntaxEditor editor, SemanticModel semanticModel, CancellationToken cancellationToken);
```

Comment:
nit: we generally like abstract methods before the normal methods so you can see what subclasses will need to impl.